### PR TITLE
fix(grid): update response from emitRowSelectionEvent to have whole record

### LINKED ISSUE #12371: Grid row selection is inconsistent with and without primaryKey
## Description  
Selection API returns different types of data depending on whether `primaryKey` is set to `igx-grid`, `igx-tree-grid`, `igx-hierarchical-grid`, etc. In the case of `primaryKey` set, I get an array of primitive values. In the case of `primaryKey` not being set, I get an array of data objects behind the grid.

 * igniteui-angular version: 15.0.0
 * browser: all

## Steps to reproduce  

1. Create an app with any grid
2. Turn row selection on
3. Log `args.newSelection` on selection changing
4. Set a `primaryKey`
5. Log `args.newSelection` on selection changing

## Result  
Inconsistent results between `primaryKey` and no `primaryKey`.

## Expected result  
Consistent results. Data members should be provided in both cases.

### LINKED ISSUE #12502: owner of rowSelectionChanging is always undefined
## Description  
When handling grid's `rowSelectionChanging` event the owner of the event arguments is always undefined.

 * igniteui-angular version: 15.0.0
 * browser: all

## Steps to reproduce  
In [this](https://stackblitz.com/edit/angular-5qggu6?file=package.json) demo I am logging the owner in `rowSelectionChanging` like this:
``` typescript
  public handleRowSelection(event: IRowSelectionEventArgs) {
    console.log(event.owner);
    this.selectedRowsCount = event.newSelection.length;
    this.selectedRowIndex = event.newSelection[0];
    this.snackbarRowCount.open();
    this.snackbar.close();
  }
```
Open the demo, change the selection of any row, and observe the console. It is always logs `undefinde`.

## Result  
`owner` is `undefined`.

## Expected result  
`owner` should point to the grid component.

### PR DESCRIPTION
**Pull request by Svetloslav15**
This PR is fixing the issue, but it is only changing the response format of the event, to always return the whole rows, not only ids. 
We should consider if it is fine, or we should refactor the whole logic around grids selection and make it consistent
regardless of whether there is a primaryKey or not.

Closes #12371
Closes #12502

### Additional information (check all that apply):
 - [x] Bug fix
 - [ ] New functionality
 - [ ] Documentation
 - [ ] Demos
 - [ ] CI/CD

### Checklist:
 - [ ] All relevant tags have been applied to this PR
 - [ ] This PR includes unit tests covering all the new code ([test guidelines](https://github.com/IgniteUI/igniteui-angular/wiki/Test-implementation-guidelines-for-Ignite-UI-for-Angular))
 - [ ] This PR includes API docs for newly added methods/properties ([api docs guidelines](https://github.com/IgniteUI/igniteui-angular/wiki/Documentation-Guidelines))
 - [ ] This PR includes `feature/README.MD` updates for the feature docs
 - [ ] This PR includes general feature table updates in the root `README.MD`
 - [ ] This PR includes `CHANGELOG.MD` updates for newly added functionality
 - [ ] This PR contains breaking changes
 - [ ] This PR includes `ng update` migrations for the breaking changes ([migrations guidelines](https://github.com/IgniteUI/igniteui-angular/wiki/Update-Migrations))
 - [ ] This PR includes behavioral changes and the feature specification has been updated with them

### CODE REVIEW

**Review by Svetloslav15 — COMMENTED** [review]
grid-virtual
When we use grid remote virtualization with scroll sample, select a row then when we scroll down and the new chunk of data is loaded, select row from it and we can see that the newSelection array is storing only the selected rows from the current chunk and currently selected rows are missing.

File: projects/igniteui-angular/src/lib/grids/tree-grid/tree-grid-selection.service.ts [hunk 1]
```diff
@@ -183,8 +183,8 @@ export class IgxTreeGridSelectionService extends IgxGridSelectionService {
      * populates the rowsToBeSelected and rowsToBeIndeterminate sets
      * with the rows which will be eventually in selected/indeterminate state
      */
-    private calculateRowsNewSelectionState(args: any) {
-        this.rowsToBeSelected = new Set<any>(args.oldSelection ? args.oldSelection : this.getSelectedRows());
+    private calculateRowsNewSelectionState(args: any, shouldConvert = false) {
+        this.rowsToBeSelected = new Set<any>(args.oldSelection ? shouldConvert ? args.oldSelection.map(r => r[this.grid.primaryKey]) : args.oldSelection : this.getSelectedRows());
```
Comment (hanastasov) [inline projects/igniteui-angular/src/lib/grids/tree-grid/tree-grid-selection.service.ts:187]: this.rowsToBeSelected will contain one undefined item when PK is not set on the grid

File: projects/igniteui-angular/src/lib/grids/tree-grid/tree-grid-selection.service.ts [hunk 2]
```diff
@@ -160,15 +159,16 @@ export class IgxTreeGridSelectionService extends IgxGridSelectionService {
      * @param visibleRowIDs list of all visible rowIds
      * @returns a new set with all direct parents of the rows within rowsToBeProcessed set
      */
-    private collectRowsChildrenAndDirectParents(rowsToBeProcessed: Set<any>, visibleRowIDs: Set<any>, adding: boolean): Set<any> {
+    private collectRowsChildrenAndDirectParents(rowsToBeProcessed: Set<any>, visibleRowIDs: Set<any>, adding: boolean, shouldConvert: boolean): Set<any> {
         const processedRowsParents = new Set<any>();
-        Array.from(rowsToBeProcessed).forEach((rowID) => {
+        Array.from(rowsToBeProcessed).forEach((row) => {
+            const rowID = shouldConvert ? row[this.grid.primaryKey] : row;
```
Comment (hanastasov) [inline projects/igniteui-angular/src/lib/grids/tree-grid/tree-grid-selection.service.ts:165]: if PK is not set this result in rowID === undefined, which is then added to selection Set and possibly causing bugs. SHould be fixed.

**Review by hanastasov — CHANGES_REQUESTED** [review]

File: projects/igniteui-angular/src/lib/grids/selection/selection.service.ts [hunk 3]
```diff
@@ -420,57 +432,100 @@ export class IgxGridSelectionService {
...
+        this.emitRowSelectionEvent(newSelection, [this.getRowDataById(rowID)], removed, event, selectedRows);
+    }
 
+    public selectPivotRowById(rowID, clearPrevSelection: boolean, event?): void {
         const selectedRows = this.getSelectedRows();
-        const newSelection = clearPrevSelection ? [rowID] : this.rowSelection.has(rowID) ?
-            selectedRows : [...selectedRows, rowID];
-        const removed = clearPrevSelection ? selectedRows : [];
-        this.emitRowSelectionEvent(newSelection, [rowID], removed, event);
+        const newSelection = clearPrevSelection ? [rowID] : this.rowSelection.has(rowID) ? selectedRows : [...selectedRows, rowID];
```
Comment (hanastasov) [inline projects/igniteui-angular/src/lib/grids/selection/selection.service.ts:448]: PivotGrid supports only single selection, so maybe this line can be simpler. I could not think of a scenario where newSelection would contain more than one row.

Anyway, for Pivot I was not able to see any selection in the event arguments:

image


I used the Pivot Grid sample.

**Review by hanastasov — CHANGES_REQUESTED** [review]

File: projects/igniteui-angular/src/lib/grids/selection/selection.service.ts (hunk 3 above)
Comment (ddincheva) [reply projects/igniteui-angular/src/lib/grids/selection/selection.service.ts:448]: PivotGridSelection

**Review by hanastasov — CHANGES_REQUESTED** [review]
1) Add this sample https://stackblitz.com/edit/angular-z9ynz5?file=src/app/grid/grid-remote-paging-defaultTemplate-sample/remote-paging-default-template.component.html with `primaryKey="ID"`

1) Select rows 2,5,6 on first page
2) Put a breakpoint in selection.service.ts, line 453
3) Go to next page and add row to selection:
image

When you go back to previous page, only row 6 is kept in selection

**Review by hanastasov — DISMISSED** [review]

**Review by MayaKirova — APPROVED** [review]

**Review by hanastasov — APPROVED** [review]

**Pull request merged**